frame/dvm/src/lib.rs

```diff
@@ -332,37 +349,154 @@ pub mod pallet {
+			return Err(InvalidTransaction::Payment.into());
+		}
+
+		let min_gas_price = T::FeeCalculator::min_gas_price();
+		if transaction.gas_price < min_gas_price {
+			return Err(InvalidTransaction::Payment.into());
+		}
+
+		Ok(account_data.nonce)
+	}
+
+	// Controls that must be performed by the pool.
+	// The controls common with the State Transition Function (STF) are in
+	// the function `validate_transaction_common`.
+	fn validate_transaction_in_pool(
+		origin: H160,
+		transaction: &TransactionV0,
+	) -> TransactionValidity {
+		let account_nonce = Self::validate_transaction_common(origin, transaction)?;
+		if transaction.nonce < account_nonce {
+			return Err(InvalidTransaction::Stale.into());
+		}
+
+		// The tag provides and requires must be filled correctly according to the nonce.
+		let mut builder = ValidTransactionBuilder::default()
+			.and_provides((origin, transaction.nonce))
+			.priority(transaction.gas_price.unique_saturated_into());
+
+		// In the context of the pool, a transaction with
+		// too high a nonce is still considered valid
+		if transaction.nonce > account_nonce {
+			if let Some(prev_nonce) = transaction.nonce.checked_sub(1.into()) {
+				builder = builder.and_requires((origin, prev_nonce))
+			}
+		}
+
+		builder.build()
+	}
+
+	/// Validate an Ethereum transaction already in block
+	///
+	/// This function must be called during the pre-dispatch phase
+	/// (just before applying the extrinsic).
+	pub fn validate_transaction_in_block(
+		origin: H160,
+		transaction: &ethereum::TransactionV0,
+	) -> Result<(), TransactionValidityError> {
+		let account_nonce = Self::validate_transaction_common(origin, transaction)?;
+
+		// In the context of the block, a transaction with a nonce that is
+		// too high should be considered invalid and make the whole block invalid.
+		if transaction.nonce > account_nonce {
+			Err(TransactionValidityError::Invalid(
+				InvalidTransaction::Future,
 			))
+		} else if transaction.nonce < account_nonce {
+			Err(TransactionValidityError::Invalid(InvalidTransaction::Stale))
+		} else {
+			Ok(())
+		}
+	}
+
+	/// Execute transaction from EthApi or PreLog Block
+	/// NOTE: For the rpc transaction, the execution will return ok(..) even when encounters error
+	/// 	  from evm runner
+	fn apply_validated_transaction(
+		source: H160,
+		transaction: TransactionV0,
+	) -> DispatchResultWithPostInfo {
+		Self::raw_transact(source, transaction.into()).map(|(_, used_gas)| {
+			Ok(PostDispatchInfo {
+				actual_weight: Some(T::GasWeightMapping::gas_to_weight(
+					used_gas.unique_saturated_into(),
+				)),
+				pays_fee: Pays::No,
+			}
 			.into())
 		})?
 	}
 
-	/// Execute DVMTransaction in evm runner and save the execution info in Pending
-	fn raw_transact(transaction: DVMTransaction) -> Result<(ExitReason, U256), DispatchError> {
+	// Execute Transaction in evm runner and save the execution info in Pending
+	fn raw_transact(
+		source: H160,
+		dvm_transaction: DVMTransaction,
```

Comment:
After we separate the source from DVMTransaction,  do we need this struct anymore? or use TransactionV0 directly?

Reply:
We can't do this, because `gas_price` should be none in the internal transact execute process.  And we can't even set the gas_price=0, this brings execution error in the evm runner.